### DISCUSSION TITLE
looking for some input for creating a stable user connection and a special channel with pusher

### DISCUSSION THREAD

**Discussion by marte1337** [neuefische/questions#610] [Web]
## 🎉 Introduction:

I´m trying to integrate a true multiplayer into my chess app, this is the according [US](https://github.com/marte1337/capstone-project/issues/22), and  this is what i got so far:

I initiated the [pusher api](https://pusher.com/docs/channels/getting_started/javascript/), so that I can create a direct connection with multiple users. I create a prelobby page where users can enter their player name and a gamelobby page where users can chat through a pusher channel. The gamelobby channel is a slightly modified version of the following example from the pusher docs:

https://pusher.com/docs/channels/getting_started/javascript-realtime-user-list/

This is how my connection works:

`lib/Index.js `: This creates the pusher-instance on the "server-side"

`api/pusher/auth/index` : This is the authorisation endpoint, where users get authorised.  The `randomString` variable is just there to make it work for now. I will need to create a more secure id in the future.

`api/pusher/index` : This is where the channels live. The` index.js` houses the chat-room of the game lobby. I already created a `presence-board.js` that should work as the private user channel that makes up the chess-game itself.

In the `prelobby.js` page the username is entered and gets send as global state to the `_app.js`, where it gets send back to all pages as prop.

The `lobby.js`is where the chatroom-channel lives, from there you can enter the chessboard-channel that is housed in the `multiplayer.js`page.

Connecting users and authorising them works. But there are a few issues, I wasn´t able to fix yet:

1.  On reload the username gets lost, but the user instance still seems to have authorisation. It´s still possible to send messages, but there is no username being shown. Also the signoff doesn´t seem to work properly: If I sign off and switch back directly to the gamelobby per URL I return with my username. There seems to be some kind of a session storage for the username somewhere, but I wasn´t able to get a grip on that one yet.

2. EDIT: Issue 2. has been solved.

3. The real challenge is the chessboard-channel: It needs to be restricted to two users, so that only the players can join it. I already created a channel, that is joinable from inside the lobby (`presence-board`, it lives inside the `multiplayer.js` page), but this one doesn´t match the criteria from above yet. It´s basicaly just another chatroom channel, that also sends chessmove-data back and forth. Playing a game is possible when two users join the channel at the same time - a page refresh or a user leaving and then reconnecting to the page will reset the game though (this might be solved with with Issue 1.)
In the final stadium, it should also be possible to create multiple two-player-channels at the same time, so that multiple games can happen. 

I would be grateful for any kind of input on those three topics.

## 🧮 Pull Request or link to code:

https://github.com/marte1337/capstone-project/pull/23

## ♻️ What are the steps to reproduce this issue?

Open multiple prelobby tabs, enter different usernames and connect. You can imitate a chat with multiple users, by sending messages from the different tabs. If you reload a tab, the user loose their username but are still signed in. Their following messages are still being displayed, but with blank usernames.

For Issue 3. go to the gamelobby and click on "Go to chessboard". It opens another chatroom without user count restriction and the refresh/reset-game situation.

## 🤖 Any logs, error output, etc?

none so far...

**Comment by ThomasF85** [comment]
Hi Marte,

regarding your questions:
1. You are managing your username state in local state (without for example persisting it in localStorage), so whenever you perform a page reload it is lost. I have not worked with the pusher api before, but it will likely persist your session via a cookie, so it persists after reloading the page.
2. You should only try to access an environment variable in the browser when it is no secret and save to share with your frontend. Otherwise you should only use it in your server side code. If you want to expose a certain environment variable you can do so by prefixing it: https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#exposing-environment-variables-to-the-browser
3. This is a pusher api specific question, so I cannot give you advice on how this works without diving into it deeply.

**Reply by marte1337** [reply]
Thank you, for the quick reply, I could fix the 2. issue with your help!!
